### PR TITLE
pipeline-manager: SQL MAP type

### PR DESCRIPTION
**Pull request by snkas**
**Outstanding tasks/questions:**
- [x] The current MAP type in pipeline-types does not take key / value type arguments inside. Is this needed?
- [x] Adapters crate: JSON and Parquet schema type correspondence
- [x] An error occurs in the compilation of the program itself (see below)

**Error in program:**
```
   Compiling project01904ed3-b496-76d3-8b29-af9a246e9671 v0.1.0 (/home/<user>/.feldera/cargo_workspace/project01904ed3-b496-76d3-8b29-af9a246e9671)
error[E0277]: the trait bound `BTreeMap<i32, i32>: DeserializeWithContext<'_, SqlSerdeConfig>` is not satisfied
   --> project01904ed3-b496-76d3-8b29-af9a246e9671/src/main.rs:116:9
    |
116 | /         deserialize_table_record!(struct_0["t", 1] {
117 | |             (field0, "map", false, Option<BTreeMap<i32, i32>>, Some(None))
118 | |         });
    | |__________^ the trait `DeserializeWithContext<'_, SqlSerdeConfig>` is not implemented for `BTreeMap<i32, i32>`, which is required by `DeserializationContext<'_, SqlSerdeConfig, std::option::Option<BTreeMap<i32, i32>>>: DeserializeSeed<'de>`
    |
    = help: the following other types implement trait `DeserializeWithContext<'de, C>`:
              <bool as DeserializeWithContext<'de, C>>
              <char as DeserializeWithContext<'de, C>>
              <isize as DeserializeWithContext<'de, C>>
              <i8 as DeserializeWithContext<'de, C>>
              <i16 as DeserializeWithContext<'de, C>>
              <i32 as DeserializeWithContext<'de, C>>
              <i64 as DeserializeWithContext<'de, C>>
              <i128 as DeserializeWithContext<'de, C>>
            and 44 others
    = note: required for `std::option::Option<BTreeMap<i32, i32>>` to implement `DeserializeWithContext<'_, SqlSerdeConfig>`
    = note: required for `DeserializationContext<'_, SqlSerdeConfig, std::option::Option<BTreeMap<i32, i32>>>` to implement `DeserializeSeed<'de>`
    = note: this error originates in the macro `deserialize_table_record` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `BTreeMap<i32, i32>: DeserializeWithContext<'_, SqlSerdeConfig>` is not satisfied
    --> project01904ed3-b496-76d3-8b29-af9a246e9671/src/main.rs:116:9
     |
116  | /         deserialize_table_record!(struct_0["t", 1] {
117  | |             (field0, "map", false, Option<BTreeMap<i32, i32>>, Some(None))
118  | |         });
     | |          ^
     | |          |
     | |__________the trait `DeserializeWithContext<'_, SqlSerdeConfig>` is not implemented for `BTreeMap<i32, i32>`, which is required by `DeserializationContext<'_, SqlSerdeConfig, std::option::Option<BTreeMap<i32, i32>>>: DeserializeSeed<'de>`
     |            required by a bound introduced by this call
     |
     = help: the following other types implement trait `DeserializeWithContext<'de, C>`:
               <bool as DeserializeWithContext<'de, C>>
               <char as DeserializeWithContext<'de, C>>
               <isize as DeserializeWithContext<'de, C>>
               <i8 as DeserializeWithContext<'de, C>>
               <i16 as DeserializeWithContext<'de, C>>
               <i32 as DeserializeWithContext<'de, C>>
               <i64 as DeserializeWithContext<'de, C>>
               <i128 as DeserializeWithContext<'de, C>>
             and 44 others
     = note: required for `std::option::Option<BTreeMap<i32, i32>>` to implement `DeserializeWithContext<'_, SqlSerdeConfig>`
     = note: required for `DeserializationContext<'_, SqlSerdeConfig, std::option::Option<BTreeMap<i32, i32>>>` to implement `DeserializeSeed<'de>`
note: required by a bound in `next_value_seed`
    --> /home/<user>/.cargo/registry/src/index.crates.io-6f17d22bba15001f/serde-1.0.203/src/de/mod.rs:1811:12
     |
1809 |     fn next_value_seed<V>(&mut self, seed: V) -> Result<V::Value, Self::Error>
     |        --------------- required by a bound in this associated function
1810 |     where
1811 |         V: DeserializeSeed<'de>;
     |            ^^^^^^^^^^^^^^^^^^^^ required by this bound in `MapAccess::next_value_seed`
     = note: this error originates in the macro `deserialize_table_record` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `BTreeMap<i32, i32>: DeserializeWithContext<'_, SqlSerdeConfig>` is not satisfied
    --> project01904ed3-b496-76d3-8b29-af9a246e9671/src/main.rs:116:9
     |
116  | /         deserialize_table_record!(struct_0["t", 1] {
117  | |             (field0, "map", false, Option<BTreeMap<i32, i32>>, Some(None))
118  | |         });
     | |          ^
     | |          |
     | |__________the trait `DeserializeWithContext<'_, SqlSerdeConfig>` is not implemented for `BTreeMap<i32, i32>`, which is required by `DeserializationContext<'_, SqlSerdeConfig, std::option::Option<BTreeMap<i32, i32>>>: DeserializeSeed<'de>`
     |            required by a bound introduced by this call
     |
     = help: the following other types implement trait `DeserializeWithContext<'de, C>`:
               <bool as DeserializeWithContext<'de, C>>
               <char as DeserializeWithContext<'de, C>>
               <isize as DeserializeWithContext<'de, C>>
               <i8 as DeserializeWithContext<'de, C>>
               <i16 as DeserializeWithContext<'de, C>>
               <i32 as DeserializeWithContext<'de, C>>
               <i64 as DeserializeWithContext<'de, C>>
               <i128 as DeserializeWithContext<'de, C>>
             and 44 others
     = note: required for `std::option::Option<BTreeMap<i32, i32>>` to implement `DeserializeWithContext<'_, SqlSerdeConfig>`
     = note: required for `DeserializationContext<'_, SqlSerdeConfig, std::option::Option<BTreeMap<i32, i32>>>` to implement `DeserializeSeed<'de>`
note: required by a bound in `next_element_seed`
    --> /home/<user>/.cargo/registry/src/index.crates.io-6f17d22bba15001f/serde-1.0.203/src/de/mod.rs:1716:12
     |
1714 |     fn next_element_seed<T>(&mut self, seed: T) -> Result<Option<T::Value>, Self::Error>
     |        ----------------- required by a bound in this associated function
1715 |     where
1716 |         T: DeserializeSeed<'de>;
     |            ^^^^^^^^^^^^^^^^^^^^ required by this bound in `SeqAccess::next_element_seed`
     = note: this error originates in the macro `deserialize_table_record` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `BTreeMap<i32, i32>: SerializeWithContext<SqlSerdeConfig>` is not satisfied
    --> project01904ed3-b496-76d3-8b29-af9a246e9671/src/main.rs:119:9
     |
119  | /         serialize_table_record!(struct_0[1]{
120  | |             field0["map"]: Option<BTreeMap<i32, i32>>
121  | |         });
     | |          ^
     | |          |
     | |__________the trait `SerializeWithContext<SqlSerdeConfig>` is not implemented for `BTreeMap<i32, i32>`, which is required by `SerializationContext<'_, SqlSerdeConfig, std::option::Option<BTreeMap<i32, i32>>>: Serialize`
     |            required by a bound introduced by this call
     |
     = help: the following other types implement trait `SerializeWithContext<C>`:
               <bool as SerializeWithContext<C>>
               <char as SerializeWithContext<C>>
               <isize as SerializeWithContext<C>>
               <i8 as SerializeWithContext<C>>
               <i16 as SerializeWithContext<C>>
               <i32 as SerializeWithContext<C>>
               <i64 as SerializeWithContext<C>>
               <i128 as SerializeWithContext<C>>
             and 35 others
     = note: required for `std::option::Option<BTreeMap<i32, i32>>` to implement `SerializeWithContext<SqlSerdeConfig>`
     = note: required for `SerializationContext<'_, SqlSerdeConfig, std::option::Option<BTreeMap<i32, i32>>>` to implement `Serialize`
note: required by a bound in `circuit::{closure#0}::_::_serde::ser::SerializeStruct::serialize_field`
    --> /home/<user>/.cargo/registry/src/index.crates.io-6f17d22bba15001f/serde-1.0.203/src/ser/mod.rs:1860:21
     |
1858 |     fn serialize_field<T>(&mut self, key: &'static str, value: &T) -> Result<(), Self::Error>
     |        --------------- required by a bound in this associated function
1859 |     where
1860 |         T: ?Sized + Serialize;
     |                     ^^^^^^^^^ required by this bound in `SerializeStruct::serialize_field`
     = note: this error originates in the macro `serialize_table_record` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
error: could not compile `project01904ed3-b496-76d3-8b29-af9a246e9671` (bin "project01904ed3-b496-76d3-8b29-af9a246e9671") due to 5 previous errors

exit code: exit status: 101
```

Is this a user-visible change (yes/no): no

### CODE REVIEW

**Comment by gz** [comment]
> The current MAP type in pipeline-types does not take key / value type arguments inside. Is this needed?
For parquet etc. we need to know the key/value type, yes

**Comment by snkas** [comment]
It seems in JSON and Parquet etc. they are inferred from `schema.fields` somehow? I see that in pipeline-types there's only one enumeration variant that has an argument added to it (INTERVAL), so is it there also needed for Map?

**Comment by gz** [comment]
It might be that it doesn't show up in the schema.json yet (best is probably to manually inspect the schema.json file and see where the key and value types show up), if they are missing @mihaibudiu needs to add them.

**Review by mihaibudiu — APPROVED** [review]

**Comment by mihaibudiu** [comment]
@ryzhyk I have added a commit which implements DeserializeWithContext for Map which needs to be reviewed.
We use BTreeMap for SQL Maps.

**Comment by ryzhyk** [comment]
> @ryzhyk I have added a commit which implements DeserializeWithContext for Map which needs to be reviewed. We use BTreeMap for SQL Maps.

Looks good,  but could use tests.

**Comment by mihaibudiu** [comment]
The added SQL test compiles, which wasn't working before.
For writing runtime tests to work the panic!() calls in the first commit have to be replaced with working code.

**Comment by mihaibudiu** [comment]
Maybe I am wrong. I guess these panics are for kafka and parquet.

**Comment by snkas** [comment]
Yes, I'm not as familiar with the format code in adapters to quickly write the implementation, hence I added the two placeholders with `panic()`. @ryzhyk Would you have time to write them?

**Comment by ryzhyk** [comment]
> Yes, I'm not as familiar with the format code in adapters to quickly write the implementation, hence I added the two placeholders with `panic()`. @ryzhyk Would you have time to write them?

Sure

**Comment by mihaibudiu** [comment]
I will try to add some tests for maps too, but not sure I will get to this today.

**Comment by ryzhyk** [comment]
I implemented support for MAP in `adapters` and added some tests. One limitation is that JSON and Avro formats will only work for maps whose key type is `STRING`. The best way to model other kinds of maps so they work with those formats is as `ARRAY<STRUCT<key STRING, value STRING>>`.

File: crates/adapters/src/test/data.rs [hunk 1]
```diff
@@ -245,8 +255,8 @@ deserialize_table_record!(EmbeddedStruct["EmbeddedStruct", 1] {
     rkyv::Serialize,
     rkyv::Deserialize,
 )]
-#[archive_attr(derive(Clone, Ord, Eq, PartialEq, PartialOrd))]
-#[archive(compare(PartialEq, PartialOrd))]
+#[archive_attr(derive(Ord, Eq, PartialEq, PartialOrd))]
+//#[archive(compare(PartialEq, PartialOrd))]
```
Comment (mihaibudiu) [inline crates/adapters/src/test/data.rs:259]: can this be deleted?

**Review by mihaibudiu — APPROVED** [review]

**Pull request merged**